src/sas/sascalc/fit/fitstate.py

```diff
@@ -0,0 +1,407 @@
+            print("="*20, "Fit page", k+1)
+            #print(fit)
+            for attr, value in sorted(fit.__dict__.items()):
+                if isinstance(value, (list, tuple)):
+                    print(attr)
+                    for item in value:
+                        print("   ", item)
+                else:
+                    print(attr, value)
+        if self.simfit:
+            print("="*20, "Constraints")
+            #print(self.simfit)
+            for attr, value in sorted(self.simfit.__dict__.items()):
+                if isinstance(value, (list, tuple)):
+                    print(attr)
+                    for item in value:
+                        print("   ", item)
+                else:
+                    print(attr, value)
+
+    def make_fitproblem(self):
+        """
+        Build collection of bumps fitness calculators and return the FitProblem.
+        """
+        # TODO: batch info not stored with project/analysis file (ticket #907)
+        models = [make_fitness(state) for state in self.fits]
+        if not models:
+            raise RuntimeError("Nothing to fit")
+        fit_problem = FitProblem(models)
+        fit_problem.setp_hook = ParameterExpressions(models)
+        return fit_problem
+
+    def _set_constraints(self):
+        """
+        Adds fit_page and constraints list to each model.
+        """
+        # early return if no sim fit
+        if self.simfit is None:
+            for fit in self.fits:
+                fit.fit_page = 'M1'
+                fit.constraints = {}
+            return
+
+        # Note: simfitpage.py:load_from_save_state relabels the model and
+        # constraint on load, replacing the model name in the constraint
+        # expression with the new name for every constraint expression. We
+        # don't bother to do that here since we don't need to relabel the
+        # model ids.
+        constraints = {}
+        for item in self.simfit.constraints_list:
+            # model_cbox in the constraints list should match fit_page_source
+            # in the model list.
+            pairs = constraints.setdefault(item['model_cbox'], [])
+            pairs.append((item['param_cbox'], item['constraint']))
+
+        # No way to uniquely map the page id (M1, M2, etc.) to the different
+        # items in fits; neither fit_number nor fit_page_source is in the
+        # pagestate for the fit, and neither model_name nor data name are
+        # unique.  The usual case of one model per data file will get us most
+        # of the way there but there will be ambiguity when the data file
+        # is not unique, e.g., when different parts of the data set are
+        # fit with different models.  If the same model and same data are
+        # used (e.g., with different background, scale or resolution in
+        # different segments) then the model-fit association will be assigned
+        # arbitrarily based on whichever happens to come first.
```

Comment:
Just want to make sure.  Will the current setup work in simultaneous fitting to multiple data sets?

Reply:
From my comment it seems there are ambiguities in the saved format, so some will fail to reload properly.  I'm guessing this is true for the 4.x GUI as well (since I presumably took the code from there in the first place).  Maybe Jeff knows the answer to this?

Are these ambiguities gone in the 5.0 save format?

Reply:
I now raise an error if the save file is ambiguous.